### PR TITLE
ORTmodule onnx input support

### PR DESCRIPTION
**Pull request by almaly-microsoft**
ORTmodule wraps pytorch model and executes training for it. However pytorch models are stored in ONNX format and in some cases (like quantization emulation) they needs to be retrained to get accuracy back.

This change is about adding that support. I updated one test to store pytorch model into ONNX format and then passed it into ORTModule. Training runs well.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 1]
```diff
@@ -81,29 +90,33 @@ def __init__(self, module):
         self._use_static_shape = False
 
         # flag to enable symbolic shape inference for dynamic shape inputs to improve performance
-        self._run_symbolic_shape_infer = True
+        self._run_symbolic_shape_infer = False
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:106]: rollback

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 2]
```diff
@@ -81,29 +90,33 @@ def __init__(self, module):
...
+                                      UserWarning)
+        else:
+            self._module_parameters = [InputParameters(i.name, inspect.Parameter.POSITIONAL_OR_KEYWORD) \
+                    for i in self._onnx_model.graph.input]
+        
 
         self.is_rocm_pytorch = (True if ((torch.version.hip is not None) and (ROCM_HOME is not None)) else False)
 
-        self._use_external_gpu_allocator = True
+        self._use_external_gpu_allocator = False
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:135]: rollback

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py [hunk 3]
```diff
@@ -40,44 +42,66 @@ class ORTModule(torch.nn.Module):
...
+        Args:
+            model: one of
+               - a PyTorch model (class that inherits from torch.nn.Module)
+               - a combined PyTorch model and loss function.
+                  Inputs to this combined PyTorch model are a concatenation of the
+                  model's input and the loss function's label input.
+                  Outputs are a concatenation of the loss function's output and the
+                  model's output.
+               - a combined ONNX model and loss function.
+            model_desc: Specify input/output shapes, types, and names.
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/ortmodule.py]: remove

File: orttraining/orttraining/python/training/ortmodule/_io.py [hunk 4]
```diff
@@ -18,44 +18,41 @@ def __init__(self,
...
         self.shape = shape
         self.require_grad_names = require_grad_names if require_grad_names else []
         self.dynamic_axes = dynamic_axes if dynamic_axes else {}
         self.schema = schema if schema else []
         self.num_positionals = num_positionals
-        self.num_positionals_non_none = num_positionals_non_none
         self.keyword_names = keyword_names
 
     def __repr__(self) -> str:
         return f'''_InputInfo class:
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/_io.py]: rollback formatting

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 5]
```diff
@@ -10,6 +10,7 @@
 from onnxruntime.capi import _pybind_state as C
 from onnxruntime.capi.onnxruntime_inference_collection import get_ort_device_type
 
+import copy
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py]: check if needed

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 6]
```diff
@@ -151,13 +158,10 @@ def backward(ctx, *grad_outputs):
 
                 backward_outputs = C.OrtValueVector()
                 self._execution_agent.run_backward(backward_inputs, backward_outputs, ctx.run_info.state)
-                # Destroy the state immediately (as opposed to be at the mercy of garbage collector) so it does not
-                # affect peak memory usage in a subsequent graph run.
-                del ctx.run_info.state
                 # Return input and initializer gradients
-                num_user_input_grads = len(self._input_info.require_grad_names)
+                num_user_input_grads = len(self._input_info.require_grad_names if self._input_info else self._graph_info.user_input_grad_names)
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py]: check if needed

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 7]
```diff
@@ -10,9 +10,10 @@
 from onnxruntime.tools.symbolic_shape_infer import SymbolicShapeInference
 
 from abc import ABC, abstractmethod
+import collections
 import copy
 import io
-import inspect
+import inspect
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: remove space

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 8]
```diff
@@ -41,15 +42,23 @@ def __init__(self, module):
...
+            self._original_module = module._original_module
+            self._flattened_module = module
+            # TODO: Single device support for now
+            self._device = _utils.get_device_from_module(module)
+        elif isinstance(module, onnx.ModelProto):
+            # Exported model
+            self._onnx_model = module
+            # Model after inference optimization or gradient building.
+            self._optimized_onnx_model = module
+            self._onnx_model_parameters = onnx_model_parameters
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: maybe you could iterate over `module.graph.initializer` and fill in this without user interaction?

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py [hunk 9]
```diff
@@ -113,8 +137,12 @@ def get_buffer(self, target: str) -> torch.Tensor:
...
 
     def parameters(self, recurse: bool = True) -> Iterator[torch.nn.Parameter]:
-        """Override original method to delegate execution to the base module"""
-        yield from self._original_module.parameters(recurse=recurse)
+        if self._onnx_model:
+            for _, parameter in self._onnx_model_parameters:
+                yield parameter
+        else:
+            """Override original method to delegate execution to the base module"""
+            yield from self._original_module.parameters(recurse=recurse)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/ortmodule.py]: you would have to override all `torch.nn.Module` APIs to prevent runtime errors

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 8 above)
Comment (almaly-microsoft) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: yep, thanks. No reason for execution manager to know about model parameters

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 5 above)
Comment (almaly-microsoft) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py]: removed

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 6 above)
Comment (almaly-microsoft) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py]: seems not needed so far

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py (hunk 9 above)
Comment (almaly-microsoft) [reply orttraining/orttraining/python/training/ortmodule/ortmodule.py]: corrected, please check
